### PR TITLE
PT-3877: Products loading reworked onto paged

### PR DESCRIPTION
**Pull request by akak1977**
## Description
Products loading reworked onto paged to avoid read of all products into the memory.
### Jira-link:
https://virtocommerce.atlassian.net/browse/PT-3876

### CODE REVIEW

File: src/VirtoCommerce.CatalogCsvImportModule.Core/Model/CsvProductMappingConfiguration.cs [hunk 1]
```diff
@@ -15,15 +15,15 @@ public CsvProductMappingConfiguration()
...
-        public string[] PropertyCsvColumns { get; set; }
+        public string[] PropertyCsvColumns { get; set; } = new string[0];
 
         public static CsvProductMappingConfiguration GetDefaultConfiguration()
         {
             var retVal = new CsvProductMappingConfiguration { Delimiter = ";" };
 
             var requiredFields = new List<string>();
             var optionalFields = ReflectionUtility.GetPropertyNames<CsvProduct>(x => x.Name, x => x.Id, x => x.Sku, x => x.CategoryPath, x => x.CategoryId, x => x.MainProductId, x => x.PrimaryImage, x => x.AltImage, x => x.SeoUrl, x => x.SeoTitle,
-                                                                                x => x.SeoDescription, x => x.SeoLanguage, x => x.SeoStore, x => x.SeoMetaKeywords, x => x.SeoImageAlternativeText, x => x.Review, x => x.ReviewType, x => x.IsActive, x => x.IsBuyable, x => x.TrackInventory,
```
Comment (ksavosteev) [inline src/VirtoCommerce.CatalogCsvImportModule.Core/Model/CsvProductMappingConfiguration.cs:26]: why delete it?

Comment (akak1977) [reply src/VirtoCommerce.CatalogCsvImportModule.Core/Model/CsvProductMappingConfiguration.cs]: I don't know. Looks like some environmental bug or I keystroked something occasionally.
Just turned it back.

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs [hunk 2]
```diff
@@ -45,89 +45,151 @@ IBlobUrlResolver blobUrlResolver
...
-                Description = "loading products..."
-            };
-
+                Description = "counting products...",
+                TotalCount = await GetProductsCount(exportInfo)
+            };
+
+            // It seems we need to read all the products twice. 
+            progressInfo.Description = "collecting product properties...";
+            progressCallback(progressInfo);
```
Comment (eMazeika) [inline src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:58]: the description changed twice, but progressCallback sent only once. Remove one of the descriptions?

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs [hunk 3]
```diff
@@ -45,89 +45,151 @@ IBlobUrlResolver blobUrlResolver
         }
 
         public async Task DoExportAsync(Stream outStream, CsvExportInfo exportInfo, Action<ExportImportProgressInfo> progressCallback)
-        {
-            var prodgressInfo = new ExportImportProgressInfo
+        {
+
```
Comment (ksavosteev) [inline src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:49]: remove line

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs [hunk 4]
```diff
@@ -45,89 +45,151 @@ IBlobUrlResolver blobUrlResolver
...
-            };
-
+                Description = "counting products...",
+                TotalCount = await GetProductsCount(exportInfo)
+            };
+
+            // It seems we need to read all the products twice. 
+            progressInfo.Description = "collecting product properties...";
+            progressCallback(progressInfo);
+            await CollectPropertyCsvColumns(exportInfo, progressCallback, progressInfo); // First time to gather all dynamic properties to have full header
```
Comment (ksavosteev) [inline src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:60]: move comment on to above line

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs [hunk 5]
```diff
@@ -45,89 +45,151 @@ IBlobUrlResolver blobUrlResolver
...
-                    prodgressInfo.ProcessedCount = counter;
-                    prodgressInfo.Description = string.Format("{0} of {1} products processed", prodgressInfo.ProcessedCount, prodgressInfo.TotalCount);
-                    if (counter % notifyProductSizeLimit == 0 || counter == prodgressInfo.TotalCount)
-                    {
-                        progressCallback(prodgressInfo);
-                    }
-                }
-            }
+            {
+
```
Comment (ksavosteev) [inline src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:72]: remove empty line

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs [hunk 6]
```diff
@@ -45,89 +45,151 @@ IBlobUrlResolver blobUrlResolver
...
+                // Fetch page by page
+                var CurrentPageNumber = 0;
+                var PageSize = 50;
```
Comment (ksavosteev) [inline src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:75]: lower case first letter

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs [hunk 7]
```diff
@@ -145,45 +207,48 @@ private List<CsvProduct> MakeMultipleExportProducts(CatalogProduct product, Dict
...
-            if (exportedProducts != null)
-            {
-                productIds = exportedProducts.ToList();
-            }
-            if (exportedCategories != null && exportedCategories.Any())
+
+        private ProductSearchCriteria ProductSearchCriteriaFactory(CsvExportInfo exportInfo)
+        {
+            ProductSearchCriteria result = null;
+            if (exportInfo.CategoryIds != null && exportInfo.CategoryIds.Any())
```
Comment (eMazeika) [inline src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:214]: use .IsNullOrEmpty()

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs [hunk 8]
```diff
@@ -145,45 +207,48 @@ private List<CsvProduct> MakeMultipleExportProducts(CatalogProduct product, Dict
...
             {
-                var criteria = new ProductSearchCriteria
+                result = new ProductSearchCriteria
                 {
-                    CatalogId = catalogId,
-                    CategoryIds = exportedCategories,
-                    Skip = 0,
-                    Take = int.MaxValue,
```
Comment (eMazeika) [inline src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:165]: add a comment why removing "Take"

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs [hunk 9]
```diff
@@ -145,45 +207,48 @@ private List<CsvProduct> MakeMultipleExportProducts(CatalogProduct product, Dict
...
+                    CategoryIds = exportInfo.CategoryIds,
                     SearchInChildren = true,
                     SearchInVariations = false,
                 };
-                var searchResult = await _productSearchService.SearchProductsAsync(criteria);
-                productIds.AddRange(searchResult.Results.Select(x => x.Id));
             }
-
-            if ((exportedCategories == null || !exportedCategories.Any()) && (exportedProducts == null || !exportedProducts.Any()))
+            if ((exportInfo.CategoryIds == null || !exportInfo.CategoryIds.Any()) && (exportInfo.ProductIds == null || !exportInfo.ProductIds.Any()))
```
Comment (eMazeika) [inline src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:224]: use .IsNullOrEmpty()
OR
can eliminate duplicating condition (exportInfo.CategoryIds == null || !exportInfo.CategoryIds.Any()?

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs [hunk 10]
```diff
@@ -145,45 +207,48 @@ private List<CsvProduct> MakeMultipleExportProducts(CatalogProduct product, Dict
...
             {
-                var criteria = new ProductSearchCriteria
+                result = new ProductSearchCriteria
                 {
-                    CatalogId = catalogId,
+                    CatalogId = exportInfo.CatalogId,
                     SearchInChildren = true,
-                    Skip = 0,
-                    Take = int.MaxValue,
```
Comment (eMazeika) [inline src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:180]: why removing "Take"?

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs [hunk 11]
```diff
@@ -145,45 +207,48 @@ private List<CsvProduct> MakeMultipleExportProducts(CatalogProduct product, Dict
...
-                productIds = searchResult.Results.Select(x => x.Id).ToList();
-            }
+            }
+
+            return result;
+        }
+
+        private async Task<int> GetProductsCount(CsvExportInfo exportInfo)
+        {
+            if (exportInfo.ProductIds != null && exportInfo.ProductIds.Length > 0)
```
Comment (eMazeika) [inline src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:239]: use .IsNullOrEmpty()

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs [hunk 12]
```diff
@@ -145,45 +207,48 @@ private List<CsvProduct> MakeMultipleExportProducts(CatalogProduct product, Dict
...
+            {
+                return exportInfo.ProductIds.Count();
+            }
+            var criteria = ProductSearchCriteriaFactory(exportInfo);
+            criteria.Skip = 0; criteria.Take = 0;
+
+            return (await _productSearchService.SearchProductsAsync(criteria)).TotalCount;
+        }
 
+        private async Task<List<CatalogProduct>> LoadProducts(List<string> productIds)
```
Comment (ksavosteev) [inline src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:250]: maybe rename to  LoadProductsWithVariations

**Review by eMazeika — APPROVED** [review]

File: src/VirtoCommerce.CatalogCsvImportModule.Core/Model/CsvProductMappingConfiguration.cs [hunk 13]
```diff
@@ -15,7 +15,7 @@ public CsvProductMappingConfiguration()
         public string Delimiter { get; set; }
         public string[] CsvColumns { get; set; }
         public ICollection<CsvProductPropertyMap> PropertyMaps { get; set; }
-        public string[] PropertyCsvColumns { get; set; }
+        public string[] PropertyCsvColumns { get; set; } = new string[0];
```
Comment (ksavosteev) [inline src/VirtoCommerce.CatalogCsvImportModule.Core/Model/CsvProductMappingConfiguration.cs:18]: use Array.Empty<string>()

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs [hunk 14]
```diff
@@ -45,89 +45,151 @@ IBlobUrlResolver blobUrlResolver
...
+                progressInfo.Description = "Done.";
+                progressCallback(progressInfo);
+            }
         }
+
+        private async Task CollectPropertyCsvColumns(CsvExportInfo exportInfo, Action<ExportImportProgressInfo> progressCallback, ExportImportProgressInfo progressInfo)
+        {
+            progressInfo.ProcessedCount = 0;
+            var CurrentPageNumber = 0;
+            var PageSize = 50;
```
Comment (ksavosteev) [inline src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:164]: lower case

**Review by ksavosteev — APPROVED** [review]

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs (hunk 8 above)
Comment (akak1977) [reply src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:165]: Skip/Take will set outside.

File: src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs (hunk 10 above)
Comment (akak1977) [reply src/VirtoCommerce.CatalogCsvImportModule.Data/Services/CsvCatalogExporter.cs:180]: Skip/Take will set outside

**Pull request merged**